doc/man/ejsmod.1
.TH EJSMOD "1" "March 2014" "ejsmod" "User Commands"
.SH NAME
ejsmod \- Ejscript Module Generator
.SH SYNOPSIS
.B ejsmod
[\fI--cslots\fR] 
[\fI--depends\fR]
[\fI--html path\fR] 
[\fI--listing\fR]
[\fI--require 'module ...'\fR]
[\fI--search ejsPath\fR]
[\fI--showDebug\fR]
[\fI--version\fR]
modules ...
.SH DESCRIPTION
The \fBejsmod\fR command is a utility program for Ejscript module files to generate symbolic assembler listings, 
slot defintion files for native classes and HTML documentation.
.SH LISTINGS
.PP
To generate symbolic assembler listings, compile your source using the \fBejsc\fR command with the --debug switch. Then
invoke \fBejsmod\fR with the \fB--listing\fR switch and supply the required modules. An assembler listing 
file with a ".lst" extension will be generated for each corresponding logical module directive. 
.PP
The listing files interleave disassembled byte code with the relevant source
code statements. The listings describe each module and its constituent functions and initializers. The 
bindings for all types and the constant pool are also listed.
.SH DOCUMENTATION
To generate HTML documentation for your code, first compile using the \fBejsc\fR command with the --doc switch. Then
invoke \fBejsmod\fR and use the \fB--html\fR switch with the directory where you want the documentation generated.
.PP
The HTML documentation will include all modules, namespaces, types, functions, properties and variable declarations.
.SH BINDING SLOT FILES
Ejscript supports the creation of native types which are coded in the C language. Native types are useful for 
core system types and for classes that must run at near binary speed. To facilitate creating native types, \fBejsmod\fR
generates slot binding definition headers which define symbolic offsets for declarations. See the Native class
samples for further information.
.SH OPTIONS
.TP
\fB\--cslots\fR
Generate slot binding definition headers for the input modules. 
.TP
\fB\--depends\fR
Generate the dependant modules required by the input modules.
.TP
\fB\--html dir\fR
Generate HTML documentation to the specified directory. The source files must have been compiled by \fBejsc\fR 
with the --doc switch to include the documentation strings in the module files.
.TP
\fB\--listing\fR
Generate symbolic assembler files for each input module.
.TP
\fB\--require 'module ...'\fR
List of modules to preload before loading the input modules.
.TP
\fB\--search ejsPath\fR
Set the module search path. The module search path is a set of directories that the \fBejsmod\fR command will use
when locating and loading Ejscript modules.  The search path will always have some system directories appended 
to the end. These include paths specified via the \fB\ EJSPATH\fR environment variable and key system directories
such as the Ejscript system module directory and the directory containing the \fBejsmod\fR command.
.IP
The search path value is similar in format to the system PATH variable format. 
On windows, path segments are separated by ";" and on Linux, Unix, FreeBSD and MAC, the path segments are separated 
by ":" delimiters.
.IP
Given a module named "a.b.c" in a script, \fBejsmod\fR will use the following search strategy to locate the module:
.IP
1. Search for a module file named "a.b.c.mod"
.IP
2. Search for a module file named "a/b/c.mod"
.IP
3. Search for a module file named "a.b.c.mod" in the search path
.IP
4. Search for a module file named c.mod in the search path
.TP
\fB\--showDebug\fR
Show debug instructions in the symbolic assembler listings.
.TP
\fB\--version\fR
Print the \fBejsmod\fR command version and exit.
.SH BACKGROUND
Ejscript is an enhanced implementation of the JavaScript language for use in embedded applications such as web 
servers, embedded devices and mobile phones. It is especially well suited for Server-Side JavaScript web applications.
.PP
Ejscript is a dynamic, interpreted, object-oriented scripting language that supports classes, objects, exceptions, 
statements, expressions and a powerful suite of data types.
.SH "REPORTING BUGS"
Report bugs to dev@embedthis.com.
.SH COPYRIGHT
Copyright \(co 2004-2014 Embedthis Software.  Ejscript is a trademark of Embedthis Software.
.br
.SH "SEE ALSO"
ejsc, ejsgen, ejs


src/cmd/ejsmod.c
/**
    ejsmod.c - Module manager 

    Copyright (c) All Rights Reserved. See details at the end of the file.
 */

/********************************** Includes **********************************/

#include    "ejsmod.h"

#if BIT_PACK_EJSCRIPT

/****************************** Forward Declarations **************************/

static void getDepends(Ejs *ejs, MprList *list, EjsModule *mp);
static void manageMod(EjsMod *mp, int flags);
static int  process(EjsMod *mp, cchar *output, int argc, char **argv);
static void require(MprList *list, cchar *name);

/************************************ Code ************************************/

MAIN(ejsmodMain, int argc, char **argv, char **envp)
{
    Mpr             *mpr;
    EjsMod          *mp;
    Ejs             *ejs;
    MprList         *requiredModules;
    char            *argp, *searchPath, *output, *modules, *name, *tok;
    int             nextArg, err, flags;

    err = 0;
    output = searchPath = 0;
    requiredModules = 0;
    
    /*
        Initialze the Multithreaded Portable Runtime (MPR)
     */
    mpr = mprCreate(argc, argv, 0);

    /*
        Allocate the primary control structure
     */
    if ((mp = mprAllocObj(EjsMod, manageMod)) == NULL) {
        return MPR_ERR_MEMORY;
    }
    mprAddRoot(mp);
    mp->lstRecords = mprCreateList(0, 0);
    mp->blocks = mprCreateList(0, 0);
    mp->docDir = sclone(".");
    mp->outputDir = sclone(".");
    
    for (nextArg = 1; nextArg < argc; nextArg++) {
        argp = argv[nextArg];
        if (*argp != '-') {
            break;
        }
        if (strcmp(argp, "--cslots") == 0) {
            mp->cslots = 1;
            mp->genSlots = 1;
            
        } else if (strcmp(argp, "--debugger") == 0 || strcmp(argp, "-D") == 0) {
            mprSetDebugMode(1);

        } else if (strcmp(argp, "--depends") == 0) {
            mp->depends = 1;

        } else if (strcmp(argp, "--dir") == 0) {
            if (nextArg >= argc) {
                err++;
            } else {
                mp->outputDir = sclone(argv[++nextArg]);
            }
            
        } else if (strcmp(argp, "--error") == 0) {
            /*
                Undocumented switch
             */
            mp->exitOnError++;
            mp->warnOnError++;
            
        } else if (strcmp(argp, "--html") == 0) {
            if (nextArg >= argc) {
                err++;
            } else {
                mp->docDir = sclone(argv[++nextArg]);
                mp->html = 1;
            }
            
        } else if (strcmp(argp, "--listing") == 0) {
            mp->listing = 1;
            mp->showAsm = 1;
            
        } else if (strcmp(argp, "--log") == 0) {
            /*
                Undocumented switch
             */
            if (nextArg >= argc) {
                err++;
            } else {
                mprStartLogging(argv[++nextArg], 0);
                mprSetCmdlineLogging(1);
            }

        } else if (strcmp(argp, "--out") == 0) {
            if (nextArg >= argc) {
                err++;
            } else {
                output = argv[++nextArg];
                mp->cslots = 1;
                mp->genSlots = 1;
            }

        } else if (strcmp(argp, "--search") == 0 || strcmp(argp, "--searchpath") == 0) {
            if (nextArg >= argc) {
                err++;
            } else {
                searchPath = argv[++nextArg];
            }

        } else if (strcmp(argp, "--version") == 0 || strcmp(argp, "-V") == 0) {
            mprPrintf("%s\n", EJS_VERSION);  
            return 0;

        } else if (strcmp(argp, "--require") == 0) {
            if (nextArg >= argc) {
                err++;
            } else {
                if (requiredModules == 0) {
                    requiredModules = mprCreateList(-1, 0);
                }
                modules = sclone(argv[++nextArg]);
                /*  Fix for Xcode and Visual Studio */
                if (modules[0] == ' ' || scmp(modules, "null") == 0) {
                    modules[0] = '\0';                    
                }
                name = stok(modules, " \t", &tok);
                while (name != NULL) {
                    require(requiredModules, name);
                    name = stok(NULL, " \t", &tok);
                }
            }

        } else if (strcmp(argp, "--warn") == 0) {
            mp->warnOnError++;

        } else if (strcmp(argp, "--xml") == 0) {
            mp->xml = 1;

        } else {
            err++;
            break;
        }
    }
    if (argc == nextArg) {
        err++;
    }
    if (mp->genSlots == 0 && mp->listing == 0 && mp->html == 0 && mp->xml == 0 && mp->depends == 0) {
        mp->listing = 1;
    }
    if (mp->depends && requiredModules == 0) {
        requiredModules = mprCreateList(-1, 0);
    }
    if (err) {
        /*
            Examples:
                ejsmod file.mod                              # Defaults to --listing
                ejsmod --listing embedthis.mod 
                ejsmod --out slots.h embedthis.mod 
         */
        mprEprintf("Usage: %s [options] modules ...\n"
            "  Ejscript module manager options:\n"
            "  --cslots              # Generate a C slot definitions file\n"
            "  --html dir            # Generate HTML documentation to the specified directory\n"
            "  --listing             # Create assembler listing files (default)\n"
            "  --out file            # Output file for all C slots (implies --cslots)\n"
            "  --require \"modules\"   # List of modules to preload\n"
            "  --search ejsPath      # Module file search path\n"
            "  --version             # Emit the program version information\n"
            "  --warn                # Warn about undocumented methods or parameters in doc\n\n", mpr->name);
        return -1;
    }

    /*
        Need an interpreter to load modules
     */
    flags = EJS_FLAG_NO_INIT;
    if (mp->html || mp->xml) {
        flags |= EJS_FLAG_DOC;
    }
    if ((ejs = ejsCreateVM(0, 0, flags)) == 0) {
        return MPR_ERR_MEMORY;
    }
    if (ejsLoadModules(ejs, searchPath, requiredModules) < 0) {
        return MPR_ERR_CANT_READ;
    }
    mp->ejs = ejs;

    if (nextArg < argc) {
        if (process(mp, output, argc - nextArg, &argv[nextArg]) < 0) {
            err++;
        }
    }
    if (mp->errorCount > 0) {
        err = -1;
    }
    mprRemoveRoot(mp);
    mp->ejs = 0;
    ejsDestroy(ejs);
    mprDestroy(MPR_EXIT_IMMEDIATE);
    return err;
}


static void manageMod(EjsMod *mp, int flags)
{
    if (flags & MPR_MANAGE_MARK) {
        mprMark(mp->ejs);
        mprMark(mp->currentLine);
        mprMark(mp->currentFileName);
        mprMark(mp->lstRecords);
        mprMark(mp->packages);
        mprMark(mp->blocks);
        mprMark(mp->currentBlock);
        mprMark(mp->docDir);
        mprMark(mp->path);
        mprMark(mp->file);
        mprMark(mp->module);
        mprMark(mp->fun);
        mprMark(mp->pc);
    }
}


static void require(MprList *list, cchar *name) 
{
    assert(list);

    if (name && *name) {
        mprAddItem(list, name);
    }
}


static int process(EjsMod *mp, cchar *output, int argc, char **argv)
{
    Ejs         *ejs;
    EjsModule   *module;
    MprFile     *outfile;
    MprList     *depends;
    int         count, i, next, moduleCount;

    ejs = mp->ejs;
    
    if (output) {
        outfile = mprOpenFile(output, O_CREAT | O_WRONLY | O_TRUNC | O_BINARY, 0664);
    } else {
        outfile = 0;
    }
    ejs->loaderCallback = (mp->listing) ? emListingLoadCallback : 0;
    mp->firstGlobal = ejsGetLength(ejs, ejs->global);

    /*
        For each module on the command line
     */
    for (i = 0; i < argc && !mp->fatalError; i++) {
        moduleCount = mprGetListLength(ejs->modules);
        ejs->loadData = mp;
        if (!mprPathExists(argv[i], R_OK)) {
            mprError("Cannot access module %s", argv[i]);
            return EJS_ERR;
        }
        if ((ejsLoadModule(ejs, ejsCreateStringFromAsc(ejs, argv[i]), -1, -1, EJS_LOADER_NO_INIT)) < 0) {
            ejs->loaderCallback = NULL;
            mprError("Cannot load module %s\n%s", argv[i], ejsGetErrorMsg(ejs, 0));
            return EJS_ERR;
        }
        if (mp->genSlots) {
            for (next = moduleCount; (module = mprGetNextItem(ejs->modules, &next)) != 0; ) {
                emCreateSlotFiles(mp, module, outfile);
            }
        }
        if (mp->depends) {
            depends = mprCreateList(-1, 0);
            for (next = moduleCount; (module = mprGetNextItem(ejs->modules, &next)) != 0; ) {
                getDepends(ejs, depends, module);
            }
            count = mprGetListLength(depends);
            for (next = 1; (module = mprGetNextItem(depends, &next)) != 0; ) {
                int version = module->version;
                mprPrintf("%@-%d.%d.%d%s", module->name, EJS_MAJOR(version), EJS_MINOR(version), EJS_PATCH(version),
                    (next >= count) ? "" : " ");
            }
            printf("\n");
        }
    }
    if (mp->html || mp->xml) {
        emCreateDoc(mp);
    }
    mprCloseFile(outfile);
    return 0;
}


static void getDepends(Ejs *ejs, MprList *list, EjsModule *mp) 
{
    EjsModule   *module;
    int         next;

    if (mprLookupItem(list, mp) < 0) {
        mprAddItem(list, mp);
    }
    for (next = 0; (module = mprGetNextItem(mp->dependencies, &next)) != 0; ) {
        if (mprLookupItem(list, module) < 0) {
            mprAddItem(list, module);
        }
    }
}

#else

MAIN(ejsMain, int argc, char **argv, char **envp)
{
    printf("Ejscript not enabled\n");
    return 255;
}

#endif /* BIT_PACK_EJSCRIPT */
/*
    @copy   default

    Copyright (c) Embedthis Software LLC, 2003-2014. All Rights Reserved.

    This software is distributed under commercial and open source licenses.
    You may use the Embedthis Open Source license or you may acquire a 
    commercial license from Embedthis Software. You agree to be fully bound
    by the terms of either license. Consult the LICENSE.md distributed with
    this software for full details and other copyrights.

    Local variables:
    tab-width: 4
    c-basic-offset: 4
    End:
    vim: sw=4 ts=4 expandtab

    @end
 */
